apps/src/lib/levelbuilder/lesson-editor/activitiesEditorRedux.js

```diff
@@ -0,0 +1,468 @@
+);
+
+export const addLevel = makeActionCreator(
+  ADD_LEVEL,
+  'activityPosition',
+  'activitySectionPosition',
+  'level'
+);
+
+export const moveActivity = makeActionCreator(
+  MOVE_ACTIVITY,
+  'activityPosition',
+  'direction'
+);
+
+export const moveActivitySection = makeActionCreator(
+  MOVE_ACTIVITY_SECTION,
+  'activityPosition',
+  'activitySectionPosition',
+  'direction'
+);
+
+export const removeActivity = makeActionCreator(
+  REMOVE_ACTIVITY,
+  'activityPosition'
+);
+
+export const removeActivitySection = makeActionCreator(
+  REMOVE_ACTIVITY_SECTION,
+  'activityPosition',
+  'activitySectionPosition'
+);
+
+export const setActivity = makeActionCreator(
+  SET_ACTIVITY,
+  'activitySectionPosition',
+  'oldActivityPosition',
+  'newActivityPosition'
+);
+
+export const addTip = makeActionCreator(
+  ADD_TIP,
+  'activityPosition',
+  'activitySectionPosition',
+  'tip'
+);
+
+function updateActivityPositions(activities) {
+  for (let i = 0; i < activities.length; i++) {
+    activities[i].position = i + 1;
+  }
+}
+
+function updateActivitySectionPositions(activities) {
+  activities.forEach(activity => {
+    let position = 1;
+    activity.activitySections.forEach(activitySection => {
+      activitySection.position = position;
+      position++;
+    });
+  });
+}
+
+function updateLevelPositions(levels) {
+  for (let i = 0; i < levels.length; i++) {
+    levels[i].position = i + 1;
+  }
+}
+
+export const NEW_LEVEL_ID = -1;
+
+function getLevels(newState, action) {
+  const activitySections =
+    newState[action.activityPosition - 1].activitySections;
+
+  return activitySections[action.activitySectionPosition - 1].levels;
+}
+
+function activities(state = [], action) {
+  let newState = _.cloneDeep(state);
+
+  switch (action.type) {
+    case INIT:
+      return action.activities;
+    case ADD_ACTIVITY: {
+      newState.push({
+        key: action.activityKey,
+        displayName: '',
+        position: action.activityPosition,
+        time: 0,
+        activitySections: []
+      });
+      updateActivityPositions(newState);
+      break;
+    }
+    case REMOVE_ACTIVITY: {
+      newState.splice(action.activityPosition - 1, 1);
+      updateActivityPositions(newState);
+      updateActivitySectionPositions(newState);
+      break;
+    }
+    case MOVE_ACTIVITY: {
+      if (
+        action.direction !== 'up' &&
+        action.activityPosition === newState.length
+      ) {
+        break;
+      }
+      const index = action.activityPosition - 1;
+      const swap = action.direction === 'up' ? index - 1 : index + 1;
+      const tempActivity = newState[index];
+      newState[index] = newState[swap];
+      newState[swap] = tempActivity;
+      updateActivityPositions(newState);
+      updateActivitySectionPositions(newState);
+      break;
+    }
+    case UPDATE_ACTIVITY_FIELD: {
+      newState[action.activityPosition - 1][action.fieldName] =
+        action.fieldValue;
+      break;
+    }
```

Comment:
This seems like a good pattern! Here and below. Thinking aloud... this doesn't let us type check the field type you are setting directly, but this is almost always going to be hooked up to a react component which will end up checking type type when it validates propTypes, so this seems ok and about as good of type checking as we have anywhere in redux.